fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -167,29 +177,201 @@ public Integer selectQueryPartitionId(@Nonnull Tuple groupKey) {
+                }
+                return oldest;
+            }
+
+            // return newest partition, unless the requested field predicate is even newer: in that case return
+            // the oldest (which guarantees no matches, and the cursor won't continue on to other partitions)
+            if ((comparisonType == GREATER_THAN || comparisonType == GREATER_THAN_OR_EQUALS) && !isAscending) {
+                LucenePartitionInfoProto.LucenePartitionInfo newest = state.context.asyncToSync(WAIT_LOAD_LUCENE_PARTITION_METADATA, getNewestPartition(groupKey, state.context, state.indexSubspace));
+                if (isNewerThan(Tuple.from(partitionFieldPredicate.getComparand()), Objects.requireNonNull(newest))) {
+                    return state.context.asyncToSync(WAIT_LOAD_LUCENE_PARTITION_METADATA, getOldestPartition(groupKey));
+                }
+                return newest;
+            }
+
+            // return comparand's "starting" partition
+            if (comparisonType == Comparisons.Type.EQUALS ||
+                    ((comparisonType == GREATER_THAN || comparisonType == GREATER_THAN_OR_EQUALS)) ||
+                    (comparisonType == LESS_THAN || comparisonType == LESS_THAN_OR_EQUALS)) {
+                return state.context.asyncToSync(WAIT_LOAD_LUCENE_PARTITION_METADATA, getStartingPartition(groupKey, Tuple.from(partitionFieldPredicate.getComparand()), isAscending, comparisonType));
+            }
         }
+
+        // default, fall-through
         return state.context.asyncToSync(WAIT_LOAD_LUCENE_PARTITION_METADATA, getNewestPartition(groupKey, state.context, state.indexSubspace));
     }
 
+    private CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> getStartingPartition(Tuple groupingKey,
+                                                                                                 Tuple partitionField,
+                                                                                                 boolean isAscending,
+                                                                                                 Comparisons.Type comparisonType) {
+        // we're here either because:
+        //  (EQUALS) or
+        //  (LESS_THAN OR LESS_THAN_OR_EQUALS) AND DESCENDING or
+        //  (GREATER_THAN_OR_EQUALS OR GREATER_THAN) AND ASCENDING
+        if ((comparisonType == GREATER_THAN || comparisonType == GREATER_THAN_OR_EQUALS) && !isAscending ||
+                (comparisonType == LESS_THAN || comparisonType == LESS_THAN_OR_EQUALS) && isAscending) {
+            throw new RecordCoreException("Internal, unexpected comparison type and sort order combination");
+        }
+
+        final Range range;
+        final int limit;
+        final boolean reverse;
+        switch (comparisonType) {
+            case GREATER_THAN_OR_EQUALS:
+                range = new Range(state.indexSubspace.subspace(groupingKey.add(PARTITION_META_SUBSPACE).addAll(partitionField)).pack(),
```

Comment:
This definitely seems more complicated than it needs to be, especially the fact that it fetches partition metadata, and then has to fetch more, or the fact that EQUALS has to iterate over all of them. Lets discuss offline after I run through some math

Reply:
basically you need to iterate in the case of EQUALS for the extreme case where you have one or more partitions where **all** the docs have the same partition field value. So you iterate until you hit the first partition that has non-equal docs too. You don't know ahead of time how many such partitions exist (super contrived in reality, but possible)